ckanext/scheming/plugins.py

```diff
@@ -234,28 +220,44 @@ def validate(self, context, data_dict, schema, action):
         thing, action_type = action.split('_')
         t = data_dict.get('type')
         if not t or t not in self._schemas:
-            return data_dict, {'type': [
-                "Unsupported dataset type: {t}".format(t=t)]}
+            return data_dict, {
+                'type': [
+                    "Unsupported dataset type: {t}".format(t=t)
+                ]
+            }
+
         scheming_schema = self._expanded_schemas[t]
 
-        if action_type == 'show':
-            get_validators = _field_output_validators
-        elif action_type == 'create':
-            get_validators = _field_create_validators
-        else:
-            get_validators = _field_validators
+        get_validators = {
+            'show': _field_output_validators,
+            'create': _field_create_validators,
+            'upate': _field_create_validators
+        }.get(action_type, _field_validators)
 
-        for f in scheming_schema['dataset_fields']:
-            schema[f['field_name']] = get_validators(
-                f,
-                scheming_schema,
-                f['field_name'] not in schema
-            )
+        fg = (
+            (scheming_schema['dataset_fields'], schema),
+            (scheming_schema['resource_fields'], schema['resources'])
+        )
 
-        resource_schema = schema['resources']
-        for f in scheming_schema.get('resource_fields', []):
-            resource_schema[f['field_name']] = get_validators(
-                f, scheming_schema, False)
+        for field_list, destination in fg:
+            for f in field_list:
+                destination[f['field_name']] = get_validators(
+                    f,
+                    scheming_schema,
+                    f['field_name'] not in schema
+                )
+
+                # Apply default field values before going through validation. This
+                # deals with fields that have form_snippet set to null, and fields
+                # that have defaults added after initial creation.
+                if data_dict.get(f['field_name']) is None:
+                    default = f.get('default')
+                    if default:
+                        data_dict[f['field_name']] = (
+                            helpers.scheming_render_from_string(
```

Comment:
I love jinja template defaults, but let's name them "default_template" instead of "default" so we still have the fast path for common non-jinja defaults.

Reply:
Which takes precedence when overloading from presets?

Reply:
let's say default_template since it's cooler. If there's a default_template in a preset but a schema wants a static value they can either use the same static value as the new default_template or set `default_template: null`